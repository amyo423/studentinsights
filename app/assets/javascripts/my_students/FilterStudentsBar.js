import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import Select from 'react-select';
import {rankedByGradeLevel} from '../helpers/SortHelpers';
import {maybeCapitalize} from '../helpers/pretty';
import {gradeText} from '../helpers/gradeText';


// Takes a list of students, uses them to find values to sort by,
// and then renders a filtering bar for different dimensions, yielding
// the sorted students to `children`.
export default class FilterStudentsBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      grade: null,
      house: null,
      counselor: null
    };

    this.onGradeChanged = this.onGradeChanged.bind(this);
    this.onHouseChanged = this.onHouseChanged.bind(this);
    this.onCounselorChanged = this.onCounselorChanged.bind(this);
  }

  filteredStudents() {
    const {students} = this.props;
    const {grade, house, counselor} = this.state;

    return students.filter(student => {
      if (shouldFilterOut(grade, student.grade)) return false;
      if (shouldFilterOut(house, student.house)) return false;
      if (shouldFilterOut(counselor, student.counselor)) return false;
      return true;
    });
  }

  onGradeChanged(grade) {
    this.setState({grade});
  }

  onHouseChanged(house) {
    this.setState({house});
  }

  onCounselorChanged(counselor) {
    this.setState({counselor});
  }

  render() {
    const {children, style, barStyle} = this.props;
    const filteredStudents = this.filteredStudents();

    return (
      <div className="FilterStudentsBar" style={style}>
        <div style={{...styles.bar, ...barStyle}}>
          <span style={styles.label}>Filter by</span>
          {this.renderGradeSelect()}
          {this.renderHouseSelect()}
          {this.renderCounselorSelect()}
        </div>
        {children(filteredStudents)}
      </div>
    );
  }

  renderGradeSelect() {
    const {students} = this.props;
    const {grade} = this.state;

    // Find all grade values in students
    const sortedGrades = _.sortBy(_.uniq(students.map(student => student.grade)), rankedByGradeLevel);
    const gradeOptions = [{value: null, label: 'All'}].concat(sortedGrades.map(grade => {
      return { value: grade, label: gradeText(grade) };
    }));
    return (
      <Select
        style={styles.select}
        simpleValue
        placeholder="Grade..."
        clearable={false}
        value={grade}
        onChange={this.onGradeChanged}
        options={gradeOptions} />
    );
  }

  renderHouseSelect() {
    const {students} = this.props;
    const {house} = this.state;

    // Find all values in students
    const sortedHouses = _.sortBy(_.uniq(_.compact(students.map(student => student.house))));
    if (sortedHouses.length === 0) return;
    const houseOptions = [{value: null, label: 'All'}].concat(sortedHouses.map(house => {
      return { value: house, label: `${maybeCapitalize(house)} house` };
    }));
    return (
      <Select
        style={styles.select}
        simpleValue
        placeholder="House..."
        clearable={false}
        value={house}
        onChange={this.onHouseChanged}
        options={houseOptions} />
    );
  }

  renderCounselorSelect() {
    const {students} = this.props;
    const {counselor} = this.state;

    // Find all values in students
    const sortedCounselors = _.sortBy(_.uniq(_.compact(students.map(student => student.counselor))));
    if (sortedCounselors.length === 0) return;
    const counselorOptions = [{value: null, label: 'All'}].concat(sortedCounselors.map(counselor => {
      return { value: counselor, label: maybeCapitalize(counselor) };
    }));
    return (
      <Select
        style={styles.select}
        simpleValue
        placeholder="Counselor..."
        clearable={false}
        value={counselor}
        onChange={this.onCounselorChanged}
        options={counselorOptions} />
    );
  }
}
FilterStudentsBar.propTypes = {
  students: PropTypes.arrayOf(PropTypes.shape({
    first_name: PropTypes.string.isRequired,
    last_name: PropTypes.string.isRequired,
    grade: PropTypes.string.isRequired,
    house: PropTypes.string,
    counselor: PropTypes.string,
  })).isRequired,
  children: PropTypes.func.isRequired,
  style: PropTypes.object,
  barStyle: PropTypes.object
};

const styles = {
  bar: {
    display: 'flex',
    alignItems: 'center'
  },
  label: {
    display: 'inline-block',
    marginBottom: 4, // fudging vertical alignment
    marginRight: 10
  },
  select: {
    display: 'inline-block',
    width: '10em',
    marginRight: 10
  }
};


function shouldFilterOut(selectedValue, studentValue) {
  if (selectedValue === null || selectedValue === undefined) return false; // no filter applied
  return (studentValue !== selectedValue);
}