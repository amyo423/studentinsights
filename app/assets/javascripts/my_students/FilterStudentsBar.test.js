import React from 'react';
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';
import _ from 'lodash';
import FilterStudentsBar from './FilterStudentsBar';
import myStudentsJson from './myStudentsJson.fixture';

function testProps(props = {}) {
  return {
    students: myStudentsJson.students,
    children: jest.fn(),
    ...props
  };
}


function testRender(props) {
  const wrapper = mount(<FilterStudentsBar {...props} />);
  return {wrapper};
}

it('renders without crashing', () => {
  const {wrapper} = testRender(testProps());
  expect(wrapper.text()).toContain('Filter by');
  expect(wrapper.text()).toContain('House...');
  expect(wrapper.text()).toContain('Grade...');
  expect(wrapper.text()).toContain('Counselor...');
});

it('filters students by grade', () => {
  const props = testProps();
  const {wrapper} = testRender(props);
  const studentsBeforeClick = props.children.mock.calls[0][0];
  expect(studentsBeforeClick.length).toEqual(myStudentsJson.students.length);
  wrapper.instance().onGradeChanged('10');

  const studentsAfterClick = props.children.mock.calls[1][0];
  expect(studentsAfterClick.length).toEqual(55);
  expect(_.uniq(studentsAfterClick.map(student => student.grade))).toEqual(['10']);
});

it('filters students by house', () => {
  const props = testProps();
  const {wrapper} = testRender(props);
  const studentsBeforeClick = props.children.mock.calls[0][0];
  expect(studentsBeforeClick.length).toEqual(myStudentsJson.students.length);
  wrapper.instance().onHouseChanged('Elm');

  const studentsAfterClick = props.children.mock.calls[1][0];
  expect(studentsAfterClick.length).toEqual(11);
  expect(_.uniq(studentsAfterClick.map(student => student.house))).toEqual(['Elm']);
});

it('filters students by counselor', () => {
  const props = testProps();
  const {wrapper} = testRender(props);
  const studentsBeforeClick = props.children.mock.calls[0][0];
  expect(studentsBeforeClick.length).toEqual(myStudentsJson.students.length);
  wrapper.instance().onCounselorChanged('WOODY');

  const studentsAfterClick = props.children.mock.calls[1][0];
  expect(studentsAfterClick.length).toEqual(22);
  expect(_.uniq(studentsAfterClick.map(student => student.counselor))).toEqual(['WOODY']);
});

it('snapshots', () => {
  const props = testProps();
  const tree = renderer
    .create(<FilterStudentsBar {...props} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
