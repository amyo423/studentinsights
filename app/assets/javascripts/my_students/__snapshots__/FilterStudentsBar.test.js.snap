// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots with all options 1`] = `
<div
  className="FilterStudentsBar EscapeListener"
  onKeyUp={[Function]}
>
  <div
    className="FilterBar"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <span
        style={
          Object {
            "display": "inline-block",
            "marginRight": 10,
          }
        }
      >
        Filter by
      </span>
      <input
        onChange={[Function]}
        placeholder="Search 74 students..."
        style={
          Object {
            "border": "1px solid #ccc",
            "borderRadius": 4,
            "display": "inline-block",
            "fontSize": 14,
            "marginLeft": 20,
            "marginRight": 10,
            "padding": "7px 7px 7px 12px",
            "width": 220,
          }
        }
        value=""
      />
      <div
        className="Select is-searchable Select--single"
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "marginLeft": 10,
              "width": "10em",
            }
          }
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-14--value"
          >
            <div
              className="Select-placeholder"
            >
              Grade...
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-14--value"
                aria-expanded="false"
                aria-haspopup="false"
                aria-owns=""
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
      <div
        className="Select is-searchable Select--single"
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "marginLeft": 10,
              "width": "10em",
            }
          }
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-15--value"
          >
            <div
              className="Select-placeholder"
            >
              House...
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-15--value"
                aria-expanded="false"
                aria-haspopup="false"
                aria-owns=""
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
      <div
        className="Select is-searchable Select--single"
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "marginLeft": 10,
              "width": "10em",
            }
          }
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-16--value"
          >
            <div
              className="Select-placeholder"
            >
              Counselor...
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-16--value"
                aria-expanded="false"
                aria-haspopup="false"
                aria-owns=""
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshots without options 1`] = `
<div
  className="FilterStudentsBar EscapeListener"
  onKeyUp={[Function]}
>
  <div
    className="FilterBar"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <span
        style={
          Object {
            "display": "inline-block",
            "marginRight": 10,
          }
        }
      >
        Filter by
      </span>
      <input
        onChange={[Function]}
        placeholder="Search 74 students..."
        style={
          Object {
            "border": "1px solid #ccc",
            "borderRadius": 4,
            "display": "inline-block",
            "fontSize": 14,
            "marginLeft": 20,
            "marginRight": 10,
            "padding": "7px 7px 7px 12px",
            "width": 220,
          }
        }
        value=""
      />
      <div
        className="Select is-searchable Select--single"
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "marginLeft": 10,
              "width": "10em",
            }
          }
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-17--value"
          >
            <div
              className="Select-placeholder"
            >
              Grade...
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-17--value"
                aria-expanded="false"
                aria-haspopup="false"
                aria-owns=""
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
